pkg/ingester/circuitbreaker.go

```diff
@@ -211,17 +214,30 @@ func (cb *circuitBreaker) tryAcquirePermit() (bool, error) {
 // that lasted longer than the configured timeout are treated as a failure.
 // The returned error is only used for testing purposes.
 func (cb *circuitBreaker) finishPushRequest(duration time.Duration, pushErr error) error {
+	return cb.finishRequest(duration, cb.cfg.PushTimeout, pushErr)
+}
+
+// finishReadRequest should be called to complete the read request executed upon a
+// successfully acquired circuit breaker permit.
+// It records the result of the push request with the circuit breaker. Read requests
+// that lasted longer than the configured timeout are treated as a failure.
+// The returned error is only used for testing purposes.
+func (cb *circuitBreaker) finishReadRequest(readDuration time.Duration, readErr error) error {
+	return cb.finishRequest(readDuration, cb.cfg.ReadTimeout, readErr)
+}
+
+func (cb *circuitBreaker) finishRequest(actualDuration time.Duration, maximalDuration time.Duration, err error) error {
 	if !cb.isActive() {
 		return nil
 	}
 	if cb.cfg.testModeEnabled {
-		duration += cb.testRequestDelay
+		actualDuration += cb.testRequestDelay
 	}
-	if cb.cfg.PushTimeout < duration {
-		pushErr = context.DeadlineExceeded
+	if maximalDuration < actualDuration {
+		err = context.DeadlineExceeded
```

Comment:
I should have posted this comment on the previous PR, but I missed this: should we check here first that `err` is not nil? Otherwise there might be a more critical error, but we're masking it under a _typical timeout_.

Reply:
I can try to reorganize the logic. I wanted to give `DeadlineExceeded` errors a higher priority because it might be that another "real" error happened during the ingester execution, but that this error is not relevant for the cb. So, it took more than the max allowed duration (which is important for the cb) to reach the part of the code where the actual error (which is not important for the cb) was returned. And if we don't prefer `DeadlineExceeded`, the cb will not be open, and we will keep executing slow parts of the code.

Reply:
Okay, can you put this as a comment on the code?

Reply:
I have changed the implementation as follows:
- in case of a relevant error for circuit breakers and a timeout, the former is preferred, a failure is recorded
- in case of an irrelevant error for circuit breakers and a timeout, the latter is preferred, a failure is recorded
- in case of no error and a timeout, the former is preferred, a failure is recorded
- in case of no error and no timeout, a success is recorded